src/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/HashAlgorithm.cs

```diff
@@ -110,6 +112,35 @@ public byte[] ComputeHash(Stream inputStream)
             return CaptureHashCodeAndReinitialize();
         }
 
+        public Task<byte[]> ComputeHashAsync(
+            Stream inputStream,
+            CancellationToken cancellationToken = default)
+        {
+            if (inputStream == null)
+                throw new ArgumentNullException(nameof(inputStream));
+            if (_disposed)
+                throw new ObjectDisposedException(null);
+
+            return ComputeHashAsyncCore(inputStream, cancellationToken);
+        }
+
+        private async Task<byte[]> ComputeHashAsyncCore(
+            Stream inputStream,
+            CancellationToken cancellationToken)
+        {
+            // Use ArrayPool.Shared instead of CryptoPool because the array is passed out.
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(4096);
+
+            int bytesRead;
+            while ((bytesRead = await inputStream.ReadAsync(buffer, 0, buffer.Length, cancellationToken).ConfigureAwait(false)) > 0)
```

Comment:
It'd be better to use the Memory-based overload of ReadAsync.

Reply:
Interesting - I'd have thought the `byte[]` overload would be preferred since the default implementation of the `Memory<byte>`-based overload uses the array pool. Is your recommendation based on that the new overload returns `ValueTask<int>` instead of `Task<int>`?

Reply:
Well, we return byte[].  Since it's an allocated reference type, allocating the Task (and being available to more API) is preferred.

Reply:
> Is your recommendation based on that the new overload returns ValueTask<int> instead of Task<int>?

Yes.  For ReadAsync in particular this is valuable because most streams will end up allocating on every ReadAsync call even when they complete synchronously.  We play a few tricks in types like MemoryStream to try to avoid doing it on every call, but it's only successful some of the time.  And some streams like NetworkStream make all Memory-based ReadAsync calls allocation-free, including ones that complete asynchronously.